airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/BasicAcceptanceTests.java

```diff
@@ -712,7 +709,141 @@ public void testResetCancelsRunningSync() throws Exception {
+
+    testHarness.assertRawDestinationContains(Collections.emptyList(), new SchemaTableNamePair("public",
+        STREAM_NAME));
+
+    // sync one more time. verify it is the equivalent of a full refresh.
+    final String expectedState = "{\"cdc\":false,\"streams\":[{\"cursor\":\"6\",\"stream_name\":\"id_and_name\",\"cursor_field\":[\"id\"],\"stream_namespace\":\"public\"}]}";
+    LOGGER.info("Starting {} sync 3", testInfo.getDisplayName());
+    final JobInfoRead connectionSyncRead3 =
+        apiClient.getConnectionApi().syncConnection(new ConnectionIdRequestBody().connectionId(connectionId));
+    waitForSuccessfulJob(apiClient.getJobsApi(), connectionSyncRead3.getJob());
+    final ConnectionState state = apiClient.getConnectionApi().getState(new ConnectionIdRequestBody().connectionId(connectionId));
+    LOGGER.info("state after sync 3: {}", state);
+
+    testHarness.assertSourceAndDestinationDbInSync(false);
+    assertEquals(Jsons.deserialize(expectedState), state.getState());
+  }
+
+  @Test
+  public void testSyncAfterUpgradeToPerStreamStateWithNoNewData(final TestInfo testInfo) throws Exception {
+    LOGGER.info("Starting {}", testInfo.getDisplayName());
+    final String connectionName = "test-connection";
+    final SourceRead source = testHarness.createPostgresSource();
+    final UUID sourceId = source.getSourceId();
+    final UUID sourceDefinitionId = source.getSourceDefinitionId();
+    final UUID destinationId = testHarness.createDestination().getDestinationId();
+    final UUID operationId = testHarness.createOperation().getOperationId();
+    final AirbyteCatalog catalog = testHarness.discoverSourceSchema(sourceId);
+
+    // Set the source to a version that does not support per-stream state
+    LOGGER.info("Setting source connector to pre-per-stream state version {}...",
+        AirbyteAcceptanceTestHarness.POSTGRES_SOURCE_LEGACY_CONNECTOR_VERSION);
+    testHarness.updateSourceDefinitionVersion(sourceDefinitionId, AirbyteAcceptanceTestHarness.POSTGRES_SOURCE_LEGACY_CONNECTOR_VERSION);
+
+    catalog.getStreams().forEach(s -> s.getConfig()
+        .syncMode(INCREMENTAL_SYNC_MODE)
+        .cursorField(List.of(COLUMN_ID))
+        .destinationSyncMode(DESTINATION_SYNC_MODE_APPEND));
+    final UUID connectionId =
+        testHarness.createConnection(connectionName, sourceId, destinationId, List.of(operationId), catalog, null).getConnectionId();
+    LOGGER.info("Beginning {} sync 1", testInfo.getDisplayName());
+
+    final JobInfoRead connectionSyncRead1 = apiClient.getConnectionApi()
+        .syncConnection(new ConnectionIdRequestBody().connectionId(connectionId));
+    waitForSuccessfulJob(apiClient.getJobsApi(), connectionSyncRead1.getJob());
+    LOGGER.info("state after sync 1: {}", apiClient.getConnectionApi().getState(new ConnectionIdRequestBody().connectionId(connectionId)));
+
+    testHarness.assertSourceAndDestinationDbInSync(false);
```

Comment:
ok here I'm actually really confused why source & destination wouldn't be in sync? shouldn't they be in sync after a sync?

Reply:
My understanding of the `false` parameter is that it is a flag that allow to check the temporary tables, but writing it like this is confusing. It is actually making sure that the source and the destination are in sync.

Reply:
Agreed that it is really confusing how that method is written. It would make it a bit more readable if we defined constants like `CHECK_SCD_TABLE = true` and `DONT_CHECK_SCD_TABLE = false`, and used those instead of passing `true` or `false` into this method directly

Reply:
Yeah...this is all throughout the test (copy and paste here).  Happy to create constants like @lmossman has suggested and update all of the tests.